client_internals_it_test.go

```diff
@@ -45,41 +45,75 @@ import (
-		testListenersAfterClientDisconnected(t, "localhost", "127.0.0.1", 46501, addListener)
-	})
-
-	t.Run("MemberHostname_ClientHostname_AddEntryListener", func(t *testing.T) {
-		testListenersAfterClientDisconnected(t, "localhost", "localhost", 47501, addEntryListener)
-	})
-	t.Run("MemberHostname_ClientHostname_AddListener", func(t *testing.T) {
-		testListenersAfterClientDisconnected(t, "localhost", "localhost", 47501, addListener)
-	})
-
-	t.Run("MemberIP_ClientIP", func(t *testing.T) {
-		testListenersAfterClientDisconnected(t, "127.0.0.1", "127.0.0.1", 48501, addEntryListener)
-	})
-	t.Run("MemberIP_ClientIP_AddListener_AddEntryListener", func(t *testing.T) {
-		testListenersAfterClientDisconnected(t, "127.0.0.1", "127.0.0.1", 48501, addListener)
-	})
+func TestClientInternal(t *testing.T) {
+	// TODO: make these tests parallized
+	testCases := []struct {
+		name string
+		f    func(t *testing.T)
+	}{
+		{name: "ClusterID", f: clientInternalClusterIDTest},
+		{name: "ClusterID_2", f: clientInternalClusterID_2Test},
+		{name: "ConnectedToMember", f: clientInternalConnectedToMemberTest},
+		{name: "EncodeData", f: clientInternalEncodeDataTest},
+		{name: "InternalInvokeOnKey", f: clientInternalInvokeOnKeyTest},
+		{name: "InternalListenersAfterClientDisconnected", f: clientInternalListenersAfterClientDisconnectedTest},
+		{name: "InvokeOnMember", f: clientInternalInvokeOnMemberTest},
+		{name: "InvokeOnPartition", f: clientInternalInvokeOnPartitionTest},
+		{name: "InvokeOnRandomTarget", f: clientInternalInvokeOnRandomTargetTest},
+		{name: "NotReceivedInvocation", f: clientInternalNotReceivedInvocationTest},
+		{name: "OrderedMembers", f: clientInternalOrderedMembersTest},
+		{name: "ProxyManagerShutdown", f: proxyManagerShutdownTest},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, tc.f)
+	}
+}
 
-	t.Run("MemberIP_ClientHostname_AddEntryListener", func(t *testing.T) {
-		testListenersAfterClientDisconnected(t, "127.0.0.1", "localhost", 49501, addEntryListener)
-	})
-	t.Run("MemberIP_ClientHostname_AddListener", func(t *testing.T) {
-		testListenersAfterClientDisconnected(t, "127.0.0.1", "localhost", 49501, addListener)
-	})
+func clientInternalListenersAfterClientDisconnectedTest(t *testing.T) {
+	testCases := []struct {
+		name   string
+		member string
+		client string
+	}{
+		{
+			name:   "MemberHostname_ClientIP",
+			member: "localhost",
+			client: "127.0.0.1",
+		},
+		{
+			name:   "MemberHostname_ClientHostname",
+			member: "localhost",
+			client: "localhost",
+		},
+		{
+			name:   "MemberIP_ClientIP",
+			member: "127.0.0.1",
+			client: "127.0.0.1",
+		},
+		{
+			name:   "MemberIP_ClientHostname",
+			member: "127.0.0.1",
+			client: "localhost",
+		},
+	}
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(fmt.Sprintf("%s_AddEntryListener", tc.name), func(t *testing.T) {
+			t.Parallel()
+			testListenersAfterClientDisconnected(t, tc.member, tc.client, addEntryListener)
```

Comment:
Why do you think that? `AddEntryListener` test has a `_AddEntryListener` the suffix and AddListener test has the `_AddListener` suffix. `testListenersAfterClientDisconnected` uses `it.NexPort()` to get the port number, so a different port number is used on every call.

Reply:
Note that the test name is used as the cluster name.

Reply:
I believe that _used to_ be the case and now it is not. I wanted to point that out

Reply:
This is `testListenersAfterClientDisconnected` which uses `t.Name()` as the cluster name and uses an automatic port:
```
func testListenersAfterClientDisconnected(t *testing.T, memberHost string, clientHost string, f func(context.Context, *hz.Map, *int64)) {
	const heartBeatSec = 6
	// launch the cluster
	port := it.NextPort()
	memberConfig := listenersAfterClientDisconnectedXMLConfig(t.Name(), memberHost, port, heartBeatSec)
	if it.SSLEnabled() {
		memberConfig = listenersAfterClientDisconnectedXMLSSLConfig(t.Name(), memberHost, port, heartBeatSec)
	}
```
It is called by the following code:
```
		t.Run(fmt.Sprintf("%s_AddEntryListener", tc.name), func(t *testing.T) {
			t.Parallel()
			testListenersAfterClientDisconnected(t, tc.member, tc.client, addEntryListener)
		})
		t.Run(fmt.Sprintf("%s_AddListener", tc.name), func(t *testing.T) {
			t.Parallel()
			testListenersAfterClientDisconnected(t, tc.member, tc.client, addListener)
		})
```
Which uses a different suffix for AddEntryListener and AddListener. Could you point me the problematic code?

Reply:
Turns out, the way the previous version worked was not ideal (using the same port numbers for adjacent cases). Thanks for the refactor!

Reply:
`listenersAfterClientDisconnectedXMLConfig` use `t.Name()` which will be resolved as same value because it takes same value  for both `AddEntryListener` and `AddListener` on the caller side. This looks problematic 🤔

Reply:
@nevzatseferoglu This is the code that calls `testListenersAfterClientDisconnected`:
```
		t.Run(fmt.Sprintf("%s_AddEntryListener", tc.name), func(t *testing.T) {
			t.Parallel()
			testListenersAfterClientDisconnected(t, tc.member, tc.client, addEntryListener)
		})
		t.Run(fmt.Sprintf("%s_AddListener", tc.name), func(t *testing.T) {
			t.Parallel()
			testListenersAfterClientDisconnected(t, tc.member, tc.client, addListener)
		})
```
A different suffix is used for AddEntryListener and AddListener, so `t.Name()` must be different. Are you sure the same test name is used? Note that, even if we don't add a suffix, Go makes the test names unique, try this example:
```
func Test1(t *testing.T) {
	t.Run("same-test", func(t *testing.T) {

	})
	t.Run("same-test", func(t *testing.T) {

	})
}
```

Reply:
`testListenersAfterClientDisconnected` takes same exact instance `t` in both side, that's why I think in that way. If testify guarantee that it will obviously unique at the end, this looks okay for me.